### PR TITLE
chores(merge): Retain owner as moderator

### LINKED ISSUE #199: Merge Components Improvements
<a href="https://github.com/mcjaeger"></img></a> **Issue by [mcjaeger](https://github.com/mcjaeger)**
_Monday Jan 08, 2018 at 09:10 GMT_
_Originally opened as https://github.com/sw360/sw360portal/issues/708_

----

When merging components in productive use, it turn actually out that quite some releases are duplicate. Then the merge of release becomes apparent ... next improvement for this functionality

Second improvement apparently needed, is that one of the creators is lost, with the effect that this person cannot write to the merged record anymore. The creator kicked out should be added to the moderators of the resulting record then.

Third thing is that if releases have (sometimes really) different release names, then it is not nice to have then entries with different release names as the resulting merged component data set. It seems to occur, when the new name is merged into the data set: then, the new resulting names does not seem to be applied to the releases of the component from which the merge assistant was triggered from.

### PR DESCRIPTION
**Pull request by lepokle**
Before this commit, one of the owner get lost during merge. Therefore one of
the owner couldn't edit the component anymore.
This commit adds functionality to automatically add the not selected owner
to the list of moderators.

Fixes #199

Signed-off-by: Leo von Klenze <leo.vonklenze@scansation.de>

### CODE REVIEW

**Comment by nutanv1contr** [comment]
Tested the PR. Not selected owner is added as moderator on the merged component. 

1. Create components by two different user.
2. Merge the components. After merge the not selected owner is added as moderator.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java [hunk 1]
```diff
@@ -530,7 +531,29 @@ private boolean isComponentUnderModeration(String componentSourceId) throws TExc
         return sourceModerationRequests.stream().anyMatch(CommonUtils::isInProgressOrPending);
     }
 
-    private void mergePlainFields(Component mergeSelection, Component mergeTarget) {
+    private void mergePlainFields(Component mergeSelection, Component mergeTarget, Component mergeSource) {
+        // First handle the creator of the component in a way, that the discarded creator will be on the 
+        // moderator list afterwards.
+        if(nullToEmpty(mergeSelection.getCreatedBy()).equals(nullToEmpty(mergeTarget.getCreatedBy()))) {
+            // creator of the target component should be retained. Add creator of source component to list of moderators.
```
[the remark this answers is not in the document]
Comment (lepokle) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: No it should not be possible, but being safe is better in my opinion.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java [hunk 2]
```diff
@@ -530,7 +531,29 @@ private boolean isComponentUnderModeration(String componentSourceId) throws TExc
...
+            mergeTarget.setModerators(mergeSelection.getModerators());
+            if(!isNullOrEmpty(mergeSource.getCreatedBy())) {
+                mergeTarget.addToModerators(mergeSource.getCreatedBy());
```
[the remark this answers is not in the document]
Comment (lepokle) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: At this place we only know that the *selected* and the target creator are the same.

But you are right: I don't know whats happening if source and target are the same. In this case no one should be added to the moderators.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (nutanv1contr) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: Correct. As far as I know there is no way we can have `createdby` field as null or empty. I would suggest to remove the null check if it is not needed. If you see any risk in removing the check then that fine.

File: backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (nutanv1contr) [reply backend/src-common/src/main/java/org/eclipse/sw360/datahandler/db/ComponentDatabaseHandler.java]: Yes correct.

**Comment by nutanv1contr** [comment]
Tested the PR again . Functionality is working as expected. 
If the source and target component `createdby` field value is different then not selected owner is added as moderator to the moderators list.

**Comment by nutanv1contr** [comment]
Code also looks good to me.

**Pull request merged**